src/java/org/apache/cassandra/index/sai/memory/MemtableIndex.java

```diff
@@ -61,32 +93,157 @@ public boolean isEmpty()
 
     public ByteBuffer getMinTerm()
     {
-        return index.getMinTerm();
+        return Arrays.stream(rangeIndexes)
+                     .map(MemoryIndex::getMinTerm)
+                     .filter(Objects::nonNull)
+                     .min(Comparator.comparing(identity(), validator))
+                     .orElse(null);
     }
 
     public ByteBuffer getMaxTerm()
     {
-        return index.getMaxTerm();
+        return Arrays.stream(rangeIndexes)
+                     .map(MemoryIndex::getMaxTerm)
+                     .filter(Objects::nonNull)
+                     .max(Comparator.comparing(identity(), validator))
+                     .orElse(null);
     }
 
-    public long index(DecoratedKey key, Clustering clustering, ByteBuffer value)
+    public void index(DecoratedKey key, Clustering clustering, ByteBuffer value, Memtable memtable, OpOrder.Group opGroup)
     {
         if (value == null || value.remaining() == 0)
-            return 0;
+            return;
 
-        long ram = index.add(key, clustering, value);
+        rangeIndexes[boundaries.getShardForKey(key)].add(key,
+                                                         clustering,
+                                                         value,
+                                                         allocatedBytes -> {
+                                                             memtable.markExtraOnHeapUsed(allocatedBytes, opGroup);
+                                                             estimatedOnHeapMemoryUsed.add(allocatedBytes);
+                                                         },
+                                                         allocatedBytes -> {
+                                                             memtable.markExtraOffHeapUsed(allocatedBytes, opGroup);
+                                                             estimatedOffHeapMemoryUsed.add(allocatedBytes);
+                                                         });
         writeCount.increment();
-        estimatedMemoryUsed.add(ram);
-        return ram;
     }
 
     public RangeIterator search(Expression expression, AbstractBounds<PartitionPosition> keyRange)
     {
-        return index.search(expression, keyRange);
+        RangeConcatIterator.Builder builder = RangeConcatIterator.builder();
+
+        for (MemoryIndex index : rangeIndexes)
+        {
+            builder.add(index.search(expression, keyRange));
+        }
```

Comment:
That's a good point. Checking the range won't affect single node clusters that much because we would be generally searching the entire key range but on multi-node clusters we are nearly always searching a specific key range.

I agree that we need to be checking here to remove shards and that we need to test single node and multi-node in-memory performance with this change.